### DISCUSSION TITLE
finding python - findpython and extensibility and uniformity

### DISCUSSION THREAD

**Discussion by bluss** [pdm-project/pdm#2014] [Ideas]
Pdm finds python through findpython. This mechanism is extensible in the sense that findpython can learn to find python from new sources.

I would like to have a framework for finding python that is more configurable, flexible and uniform. The current downside with findpython is that it can be a bit haphazard - you might find 3.10 from system python but 3.9 from a pyenv install. You'd want to ensure uniformity to make it easier for you - for example to make sure all your python versions in use are installed the same way.

Let's say you have different strategies:

- System python
- Pyenv
- Other python installers: py, asdf, rye

You would like to be able to configure: all your projects get their python from pyenv (asdf, rye, etc..) by default. In this configuration, pdm or findpython's mechanism only considers python versions from the configured source.

Some of the sources have the ability to install python on demand (pyenv, asdf, rye), even better then if we had a hook so that it would just fetch python using the configured source, if it was missing.

**Comment by frostming** [comment]
The library is easy to extend, just create a subclass of `Provider` and add it to the `Finder` object.

Here is an example in PDM:
https://github.com/pdm-project/pdm/blob/4618fc8445df1a854565d98ddf4e54771c934e9b/src/pdm/project/core.py#L660-L668

You can also modify the list of default providers by changing `findpython.finder.ALL_PROVIDERS`

**Reply by bluss** [reply]
Great. Would you accept PRs in findpython for   1) An option to set which provider(s) to use  2) A rye provider?  :slightly_smiling_face: 

Then one could look at how to maybe configure and use them from PDM

**Reply by frostming** [reply]
would accept 1) but we need to discuss the design first, after that is done, rye provider should be easy to support in a plugin

**Reply by bluss** [reply]
I wrote a RyeProvider while this discussion was happening already, would you accept it into findpython proper? Does findpython support plugins? Or is maybe there's a good reason to make it a pdm plugin instead, if there is more integration with rye that can be done(?)

**Reply by frostming** [reply]
The name "RyeProvider" would not be accurate, since the python provider has nothing to do with the package manager. It is powered by [python-build-standalone](https://github.com/indygreg/python-build-standalone). I would suggest something like `build standalone provider` though it is a bit long-winded.

**Reply by bluss** [reply]
I was thinking it will find python installed by rye. How they were built doesn't matter so much. They'll be installed into for example `~/.rye/py/cpython@3.10.11/` and that's how we find them. python-build-standalone builds could of course be installed by some other tool to some other place (let's imagine if asdf used the same builds).

**Reply by frostming** [reply]
Oh, I thought it not only finds installed Pythons, but also installs from the build-standalone if not available. In that way we don't depend on rye at all.

**Reply by bluss** [reply]
Aha that explains the difference. That's a different project than what I had in mind, I'll admit.

I'm thinking, pyenv, asdf (and rye) can all install python if you want to. To reimplement or implement an alternative to them would be a different project in itself.

When I said RyeProvider, to start with it's just something that's extending very straightforwardly, what AsdfProvider for example is doing, onto Rye.

**Reply by bluss** [reply]
The suggestion for the previously mentioned (1) part, selecting between available providers, was posted here: https://github.com/frostming/findpython/issues/17
